adapters/tlsio_openssl.c

```diff
@@ -799,11 +807,11 @@ static int decode_ssl_received_bytes(TLS_IO_INSTANCE* tls_io_instance)
 
     int rcv_bytes = 1;
 
-    while (rcv_bytes > 0)
+    while (rcv_bytes > 0 && tls_io_instance->tlsio_state == TLSIO_STATE_OPEN)
```

Comment:
would this prevent decode_ssl_recevied_bytes() from processing data that has been received before closing but not processed yet?

Reply:
I don't think so. The problem was that the tlsio_state can be changed from the callback called in the block below, and the ssl instance is closed, and then this function tries to read from it and creates an error.